service/collector/endpoint.go

```diff
@@ -47,23 +62,211 @@ func NewEndpoint(config EndpointConfig) (*Endpoint, error) {
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			ipList := getEndpointIps(endpoint)
+			for _, ip := range ipList {
+				ingressCheckState, proxyProtocol := e.ingressEndpointUp(ip, ingresSchemeHttp, ingressPortHttp)
+
+				// send ingress endpoint status metric
+				ch <- prometheus.MustNewConstMetric(
+					endpointLabelsDesc,
+					prometheus.GaugeValue,
+					ingressCheckState,
+					clusterID,
+					ip,
+					ingresSchemeHttp,
+					proxyProtocol,
+				)
+			}
+			return nil
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		return microerror.Mask(err)
+	}
 
 	return nil
 }
 
-func (i *Endpoint) Describe(ch chan<- *prometheus.Desc) error {
+func (e *Endpoint) Describe(ch chan<- *prometheus.Desc) error {
 	ch <- endpointLabelsDesc
 	return nil
 }
+
+func getEndpointIps(endpoint *corev1.Endpoints) []string {
+	var ipList []string
+	for _, subset := range endpoint.Subsets {
+		for _, address := range subset.Addresses {
+			ipList = append(ipList, address.IP)
+		}
+	}
+
+	return ipList
+}
+
+func (e *Endpoint) buildHttpRequest(ipAddress string, scheme string, port int) (*http.Request, error) {
+	u := url.URL{
+		Host:   fmt.Sprintf("%s:%d", ipAddress, port),
+		Path:   "healthz",
+		Scheme: scheme,
+	}
+
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		return nil, microerror.Maskf(err, "unable to construct health check request")
+	}
+
+	// close connection after health check request (the TCP connection gets
+	// closed by deferred s.tr.CloseIdleConnections()).
+	req.Header.Add("Connection", "close")
+
+	return req, nil
+}
+
+// ingressEndpointUp checks if ingress endpoint is up and responding to http traffic
+func (e *Endpoint) ingressEndpointUp(ipAddress string, scheme string, port int) (float64, string) {
+	err := e.ingressEndpointUpClassicHttp(ipAddress, ingresSchemeHttp, ingressPortHttp)
+
+	if err == nil {
+		// ingress endpoint check was successful, no proxy-protocol
+		return ingressCheckSucesfull, proxyProtocolFalse
+
+	}
+	// ingress endpoint check error, but we might just hit proxy protocol so try check with proxy-protocol enabled
+	err = e.ingressEndpointUpProxyProtocol(ipAddress, ingresSchemeHttp, ingressPortHttp)
+
+	if err == nil {
+		// ingress endpoint check was successful, proxy-protocol enabled
+		return ingressCheckSucesfull, proxyProtocolTrue
+	} else {
+		// ingress endpoint check failure
+		return ingressCheckFailure, proxyProtocolUnknown
+	}
+}
+
+// ingressEndpointUpClassicHttp send http packet to the endpoint to ensure target ingress endpoint ip is up
+func (e *Endpoint) ingressEndpointUpClassicHttp(ipAddress string, scheme string, port int) error {
+	req, err := e.buildHttpRequest(ipAddress, scheme, port)
+	if err != nil {
+		// failed to build http req
+		return microerror.Mask(err)
+	}
+	// be sure to close idle connection after health check is finished
+	defer e.httpTransport.CloseIdleConnections()
+
+	// send request to http endpoint
+	_, err = e.httpClient.Do(req)
```

Comment:
Response body should be closed - otherwise the connection won't be closed:

```
res, err := e.httpClient.Do(req)
if err != nil {
	return microerror.Mask(err)
}
defer res.Body.Close()
```

I'm uncertain what `e.httpTransport.CloseIdleConnections()` does. Maybe that?